core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java

```diff
@@ -71,58 +64,56 @@ public ListBucketResult() {}
+   * @param children a list of {@link URIStatus}, representing the objects and common prefixes
    * @param options the list bucket options
    */
   public ListBucketResult(
-      String bucketName, List<URIStatus> objectsList, ListBucketOptions options) {
+      String bucketName, List<URIStatus> children, ListBucketOptions options) {
     mName = bucketName;
+    mMaxKeys = DEFAULT_MAX_KEYS;
+
     mPrefix = options.getPrefix();
-    mKeyCount = 0;
-    if (options.getMaxKeys() != null) {
-      mMaxKeys = Integer.parseInt(options.getMaxKeys());
-    }
-    mContents = new ArrayList<>();
-    mContinuationToken = options.getContinuationToken();
-
-    Collections.sort(objectsList, new URIStatusComparator());
-
-    int startIndex = 0;
-    if (options.getContinuationToken() != null) {
-      URIStatus tokenStatus = new URIStatus(new FileInfo().setPath(
-          mName + AlluxioURI.SEPARATOR + mContinuationToken));
-      startIndex = Collections.binarySearch(objectsList, tokenStatus, new URIStatusComparator());
-      if (startIndex < 0) {
-        // If continuation token does not exist in the object list, find the first element which is
-        // greater than the token.
-        startIndex = (-1) * startIndex - 1;
-      }
+    mMarker = options.getMarker();
+
+    Collections.sort(children, new URIStatusComparator());
+
+    final List<URIStatus> keys = children.stream()
+        .filter((status) -> status.getPath().compareTo(mMarker) > 0)
+        .limit(mMaxKeys)
+        .collect(Collectors.toList());
+
+    mKeyCount = keys.size();
+    mIsTruncated = mKeyCount == mMaxKeys;
+
+    if (mIsTruncated) {
+      mNextMarker = keys.get(keys.size() - 1).getPath();
     }
 
-    for (int i = startIndex; i < objectsList.size(); i++) {
+    final Map<Boolean, List<URIStatus>> typeToStatus = keys.stream()
+        .collect(Collectors.groupingBy((status) -> status.isFolder()));
+    final List<URIStatus> objectsList = typeToStatus.getOrDefault(false, new ArrayList<>());
+    final List<URIStatus> prefixList = typeToStatus.getOrDefault(true, new ArrayList<>());
+
+    mContents = new ArrayList<>();
+    for (int i = 0; i < objectsList.size(); i++) {
       URIStatus status = objectsList.get(i);
-      String objectKey = status.getPath().substring(mName.length() + 1);
-      if (mKeyCount >= mMaxKeys) {
-        mIsTruncated = true;
-        mNextContinuationToken = objectKey;
-        return;
-      }
-      if (mContinuationToken != null && objectKey.compareTo(mContinuationToken) < 0) {
-        continue;
-      }
-      if (mPrefix != null && !objectKey.startsWith(mPrefix)) {
-        continue;
-      }
-      // TODO(chaomin): set ETag once there's a way to get MD5 hash of an Alluxio file.
-      // TODO(chaomin): construct the response with CommonPrefixes when delimiter support is added.
+
       mContents.add(new Content(
-          status.getPath().substring(mName.length() + 1),
+          status.getPath().substring(mName.length() + 2),
```

Comment:
why is this + 2 and not + 1 as before?

Reply:
because +1 was wrong and you got an extra character in your S3Browser result :/

Reply:
In particular, this needs to remove 2 slashes. The both ends of "/<mName>/"